### PR TITLE
upgrade xterm to fix terminal dragging between areas

### LINKED ISSUE #3838: [terminal] Not useable after dragging into a different tab folder
It seems the terminal sometimes doesn't draw any updates after dragging into a different tab folder.

### PR DESCRIPTION
**Pull request by akosyakov**
fix #3838

### CODE REVIEW

**Comment by akosyakov** [comment]
@marcdumais-work What is the process now to get CQs for runtime dependencies? This fix is quite important moving terminals around is basically broken on master.

**Comment by akosyakov** [comment]
@svenefftinge Does it address #4319?

**Comment by AlexTugarev** [comment]
Moving terminals around seem to work much better, @akosyakov. 

Have you noticed errors on the console?

Screen Shot 2019-03-25 at 08 55 53
Screen Shot 2019-03-25 at 08 56 06
Screen Shot 2019-03-25 at 08 56 13

**Comment by akosyakov** [comment]
@AlexTugarev It seem to be https://github.com/xtermjs/xterm.js/issues/1932 in xtermjs. Do you think it should be fixed before we upgrade?

**Comment by AlexTugarev** [comment]
> Do you think it should be fixed before we upgrade?

Personally, I'd say this improves the terminal a lot and it looks like there are just errors on the JS console. Despite of still broken reflow in some cases it, I haven't seen any issues using it. 

👍

**Comment by akosyakov** [comment]
From the user perspective it does not work so bad, logs get really polluted. @svenefftinge suggested to have a filter for this error for time being, @AlexTugarev sounds good?

**Comment by lmcbout** [comment]
@akosyakov 
This version allows us to use the terminal after we move it. 

If I move the terminal to the right tab bar panel, then I resize the terminal and I restart the front-end:
--> I cannot move the terminal elsewhere anymore
--> The data in the terminal are not adjusted properly until I press "enter"
--> I cannot delete the  terminal from the right tab panel unless I select "Close " from the menu
Tested on Ubuntu 16.04

See  attachment:

terminalIssue

**Comment by vince-fugnitto** [comment]
> @akosyakov
> This version allows us to use the terminal after we move it.
> 
> If I move the terminal to the right tab bar panel, then I resize the terminal and I restart the front-end:
> --> I cannot move the terminal elsewhere anymore
> --> The data in the terminal are not adjusted properly until I press "enter"
> --> I cannot delete the terminal from the right tab panel unless I select "Close " from the menu
> Tested on Ubuntu 16.04
> 
> See attachment:
> 
> terminalIssue

I don't see the same behaviour as you @lmcbout, everything works correctly for me, maybe you can show me in person later?

**Comment by marcdumais-work** [comment]
> This fix is quite important moving terminals around is basically broken on master.
+1

> @marcdumais-work What is the process now to get CQs for runtime dependencies? 

This here is a good example case, that should be very easy to deal-with, just being an updated (so previously analysed/permitted) dependency. 

No official change yet to the process has been announced:  we have one "NPM production dependencies CQ" for this whole repo and in theory should update it for new stuff. If we have only 1 PR waiting for the CQ, this works. But with multiple, we either need to wait until the CQ is resolved and then deal with the next one, and so on... Or somehow merge all those waiting PRs together and so obtain a single NPM "state" to submit in the CQ. 

As we have seen, it can take a while for those CQs to be approved, so having a single one scales poorly. Having a "CQ branch" seems like a lot of work for not much gain. 

I think we should rather have as many such CQs as needed, one per PR introducing a new or updated production dependency, each being independent from the others. Then when we submit e.g. the CQ for this PR here, the emphasis (for the Foundation) should be to check the differences vs latest approved CQ (i.e. master or close). IP issues that are applicable at large and not PR-specific (e.g. Electron) should be handled separately and not block the other CQs. 

WDYT?

**Comment by akosyakov** [comment]
@svenefftinge  @marcdumais-work Can we discuss tomorrow on dev meeting? Do we need a CQ now or we can open later and merger a PR now?

**Comment by akosyakov** [comment]
> I cannot delete the terminal from the right tab panel unless I select "Close " from the menu
Tested on Ubuntu 16.04

I believe it is expected and not relevant to this PR.

**Comment by akosyakov** [comment]
> The data in the terminal are not adjusted properly until I press "enter"

It was like that always one can drag only using the caption in sidebar, not a toolbar. If you want to allow dragging with the toolbar please file an issue. It is not related to changes in this PR.

**Comment by akosyakov** [comment]
> The data in the terminal are not adjusted properly until I press "enter"

@lmcbout I cannot reproduce it. Could you specify what you type?

**Comment by lmcbout** [comment]
@akosyakov 
- Delete the terminal from the side panel: need to use the icon and the menu: ok Just I didn't see the "X" on the title bar as I was expecting 
- Moving: Again I did not expect that I had to use the icon in the side panel to allow moving: OK  as it is 
- For the text: I just type an invalid command "xxxxxx" -> enter -> move the terminal on the right side panel ->  resize the right side panel -> restart the client 
===> the data in the terminal looks weird until I press "ENTER" to get the prompt back, then it looks ok after

**Comment by akosyakov** [comment]
This reflow feature does not feel stable, maybe we wait till it is sorted out in xterm and leave with terminals only in the bottom panel. Or help to fix it in xterm.

**Comment by lmcbout** [comment]
@akosyakov  May be the re-flow is missing, but I think we should consider to uplift xterm to the latest version, at least if someone moves it elsewhere, the terminal won't be frozen .

**Comment by marcdumais-work** [comment]
> @svenefftinge @marcdumais-work Can we discuss tomorrow on dev meeting?Do we need a CQ now or we can open later and merger a PR now? I don't want to open any CQs before we agree on something to avoid confusion.

@akosyakov Let me ping Sharon about this.

**Comment by akosyakov** [comment]
@Tyriar pointed out that bogus output on resize it is not xterm issue, but result produced by used shell: https://github.com/xtermjs/xterm.js/pull/2002#issuecomment-480866997

Since https://github.com/theia-ide/theia/pull/4680#issuecomment-476090542 is going to be fixed in 3.13.0. I suggest we wait for it and then rebase and merge this PR. I hope that 3.13.0 is already around the corner :)

**Comment by akosyakov** [comment]
3.13.0 was released, good to merge?

File: packages/terminal/package.json [hunk 1]
```diff
@@ -8,7 +8,7 @@
     "@theia/filesystem": "^0.6.0",
     "@theia/process": "^0.6.0",
     "@theia/workspace": "^0.6.0",
-    "xterm": "3.9.2"
+    "xterm": "3.13.0"
```
Comment (akosyakov) [inline packages/terminal/package.json]: @marcdumais-work Is it fine?

Comment (marcdumais-work) [reply packages/terminal/package.json]: Hi @akosyakov , let me run the ["ECD Theia Intellectual Property Clearance, Approach"](https://github.com/theia-ide/theia/wiki/Registering-CQs#case-new-or-updated-npm-production-dependency) on this PR branch and get back to you. ETA ~30 mins.

Comment (marcdumais-work) [reply packages/terminal/package.json]: Done - looks good - you can merge from a license PoV. 

There was one new extra "NOASSERTION" dependency (nugget 2.0.1), but it looks to be BSD-license after investigation:

image

**Pull request merged**